dist/common/scripts/scylla_util.py

```diff
@@ -315,10 +329,17 @@ def apt_install(pkg):
     if is_offline():
         pkg_error_exit(pkg)
     if not apt_is_updated():
-        run('apt-get update', shell=True, check=True)
+        # The lock for update and install/remove are different, and
+        # DPkg::Lock::Timeout will only wait for install/remove lock.
+        # So we have to implement wait_for_lock for update.
+        while True:
+            wait_for_lock('/var/lib/apt/lists/lock')
```

Comment:
as an alternative of checking the flock'ed file, how about just checking the output of `apt-get`? and wait until either the command finishes without error, or unexpected error shows up. as apt-pkg prints out details error in this case. see https://github.com/Debian/apt/blob/8ab8ad65ea37261b96ee0a4fbcfade19075f8629/apt-pkg/contrib/fileutl.cc#L313-L322

Reply:
Right, it also can work.
I guess it wil be something like this:
```
+        while True:
+            res = run('apt-get update', shell=True, check=False, capture_output=True, encoding='utf-8')
+            print(res.stdout)
+            print(res.stderr)
+            if res.returncode == 0:
+                break
+            else:
+                match = re.match('^E: Could not get lock ', res.stderr, re.MULTILINE)
+                # if error is "Could not get lock", wait a while and retry
+                if match:
+
+                    time.sleep(30)
+                # if error is not "Could not get lock", raise Exception
+                else:
+                    res.check_returncode()
```
Do you think this is better than current patch?

Reply:
@syuu1228 actually, i was playing around with `fcntl` in Python to get the process's pid and its command line. but that's more convoluted than your approach: we need to use `struct.pack(hhQQi, fcntl.F_WRLCK, os.SEEK_SET, 0, 0, -1)` for the `arg` parameter passed to `fcntl.fcntl(fd, fcntl.F_GETLK, arg)`. and then use `struct.unpack()` for the pid of the process holding the flock, and read `f"/proc/{pid}/status"` for the command line. it's more straightforward at the expense of exploring the implementation details of `apt`.

on one hard, i can feel the concern of Israel that we might be at the risk of going to far into a tool's internals, on the other hand, i think we should do our best to automate the process. so i agree with you that, like a lot of softwares, we can also use the backoff retry policy.

the downside of your change is, IMHO, that it hurts the maintainability. so i think the change above is better. just two nit:

1. we could have a timeout
2. no need to have an `else` branch -- we can put the error handling right at the same indent level of `if res.returncode == 0`.
3. pass `check=True` to `run()`, and handle the `CalledProcessError` exception: the stdout and stderr are included in it. simpler this way.

Reply:
Pushed new version which based on https://github.com/scylladb/scylladb/pull/16561#discussion_r1437027247.